Skipraw

Nikoh77 opened the pull request:
Like title this mod add the possibility to skip/exclude raw  file from scan - backend related code

Nikoh77 committed:
first commit for raw skipping

Nikoh77 committed:
wip raw skipping

Nikoh77 committed:
wip raw skipping

derneuere commented:
Thank you for opening this pull request! 

Can you please write a couple of django tests to confirm, that this works and to prevent future regression? :) 

You have to run the tests inside the container this time with docker exec -it librephotos python manage.py test

Nikoh77 commented:
Ok, I've never done any test but I'll try to learn

Nikoh77 committed:
"Add tests for SKIP_RAW_FILES feature

- Add 9 tests to verify RAW file skipping functionality
- Test walk_directory() and walk_files() skip RAW files when enabled
- Test cleanup of existing RAW photos when feature is enabled
- Test file type detection functions (is_raw, is_metadata)
- All tests passing with improved coverage"

Nikoh77 commented:
Ok done

sickelap commented on api/models/photo.py:
```diff
@@ -506,11 +506,36 @@ def _add_to_album_thing(self):
+                should_remove = True
+            # Check if RAW files should be skipped based on current settings
+            elif site_config.SKIP_RAW_FILES and is_raw(file.path):
+                should_remove = True
```
Is it not enough just to skip raw files? Some ppl might want to keep them. Perhaps adding an option wether remove them would be good idea?

Nikoh77 replied:
Sorry, I don't understand what you mean. This part of the code is used when you already have raw photos in the database. Activating skipraw deletes them from the database.

sickelap replied:
Yea, I read the test case and now it make sense.

derneuere commented:
I double checked. Your code will delete all missing photos, not just the raw ones. That should not happen.

```
class MissingPhotoTest(TestCase):
    """Test suite to verify that missing non-RAW photos are NOT incorrectly marked as removed."""
    
    def setUp(self):
        """Set up test user for each test."""
        self.user = create_test_user()

    def test_missing_jpeg_should_not_mark_photo_as_removed(self):
        """
        Test that when a regular JPEG file goes missing from disk, 
        the photo should NOT be marked as removed (only RAW files should trigger removal).
        
        This test demonstrates the bug where _check_files() incorrectly marks
        all missing photos as removed, not just RAW files.
        """
        with Patcher() as patcher:
            # Create a JPEG file on the filesystem
            jpeg_path = "/data/photos/vacation.jpg"
            patcher.fs.create_file(jpeg_path, contents=b"fake jpeg content")
            
            # Create a Photo with the JPEG file (simulating a previously scanned photo)
            jpeg_file = File.create(jpeg_path, self.user)
            photo = Photo.objects.create(
                image_hash=jpeg_file.hash,
                owner=self.user,
                main_file=jpeg_file,
                added_on=timezone.now()
            )
            photo.files.add(jpeg_file)
            photo.save()
            
            # Verify photo is not removed initially
            self.assertFalse(photo.removed)
            self.assertEqual(photo.files.count(), 1)
            
            # Now simulate the file being deleted from disk (e.g., user deleted it manually)
            patcher.fs.remove(jpeg_path)
            
            # Trigger _check_files() (this happens during rescan)
            photo._check_files()
            
            # Refresh from database
            photo.refresh_from_db()
            jpeg_file.refresh_from_db()
            
            # The file should be marked as missing
            self.assertTrue(jpeg_file.missing)
            
            # The file should be removed from the photo's files
            self.assertEqual(photo.files.count(), 0)
            
            # BUG: The photo should NOT be marked as removed
            # Only RAW files that are being skipped should trigger photo removal
            # Regular missing photos should remain in the database for recovery
            self.assertFalse(
                photo.removed, 
                "Missing JPEG photos should NOT be marked as removed. "
                "Only RAW files being skipped due to SKIP_RAW_FILES should trigger removal."
            )

    def test_missing_jpeg_with_metadata_should_not_mark_photo_as_removed(self):
        """
        Test that when a JPEG file goes missing but metadata remains,
        the photo should NOT be marked as removed.
        """
        with Patcher() as patcher:
            # Create a JPEG file and XMP metadata file
            jpeg_path = "/data/photos/sunset.jpg"
            xmp_path = "/data/photos/sunset.xmp"
            
            patcher.fs.create_file(jpeg_path, contents=b"fake jpeg content")
            patcher.fs.create_file(xmp_path, contents=b"fake xmp content")
            
            # Create a Photo with both files
            jpeg_file = File.create(jpeg_path, self.user)
            xmp_file = File.create(xmp_path, self.user)
            
            photo = Photo.objects.create(
                image_hash=jpeg_file.hash,
                owner=self.user,
                main_file=jpeg_file,
                added_on=timezone.now()
            )
            photo.files.add(jpeg_file)
            photo.files.add(xmp_file)
            photo.save()
            
            # Verify initial state
            self.assertFalse(photo.removed)
            self.assertEqual(photo.files.count(), 2)
            
            # Simulate JPEG being deleted but XMP remaining
            patcher.fs.remove(jpeg_path)
            
            # Trigger _check_files()
            photo._check_files()
            
            # Refresh from database
            photo.refresh_from_db()
            jpeg_file.refresh_from_db()
            
            # The JPEG file should be marked as missing
            self.assertTrue(jpeg_file.missing)
            
            # Only XMP should remain
            self.assertEqual(photo.files.count(), 1)
            self.assertIn(xmp_file, photo.files.all())
            
            # BUG: Even with only metadata remaining, if the original file was a regular photo
            # (not a RAW being skipped), the photo should NOT be marked as removed
            self.assertFalse(
                photo.removed,
                "Photo with missing JPEG but remaining metadata should NOT be marked as removed. "
                "This differs from RAW files which should be removed when skipped."
            )
```

Nikoh77 commented:
First of all, sorry for my not-so-perfect English — I’m using a translator, but I guess you already noticed that 😅.
Also, sorry if I don’t really know how to properly contribute to an open source project — this is my first time doing it.

Now, getting to the point:
I think LibrePhotos creates a photo object for every image because it computes a hash. So for example, img01.jpg would be a different photo object than img01.dng.
If I understood correctly, each of these becomes the “main” of its own photo object, and any matching .xmp file (like img01.xmp) gets assigned to the first matching image object.

If that’s true, I could end up in the following situation (which actually happened to me):
My database is full of photos of all types (raw, bitmap, and xmp). Since almost every photo appears twice — because in my workflow (like many other people I know) I keep raw files next to the jpgs — I want the raw images to stay hidden. With LibrePhotos I just want a gallery of my final, edited photos that I can show and share. When I need to look at the raw files, I use something like Darktable.

So, I enable the “skip raw” option and run a manual scan.

Here’s the core of the issue: my modification (barring any bugs) deletes all photo objects that no longer have a main file, and also those whose main file is an XMP.
In my tests, I noticed some images still showing up in the gallery just because their main raw file was removed, but the XMP file was still there.

In my opinion, that’s a mistake — LibrePhotos is an image viewer, and metadata should be secondary. If a photo object loses its actual image, it shouldn’t represent a photo anymore.

That’s what I understood from testing and reading the code. Of course, I don’t have the full picture, so I’d really appreciate your help figuring out where I might be wrong.

derneuere commented:
No problem, ChatGPT also helps me out a lot when writing messages 😄 

## How LibrePhotos is designed to behave

- **Photo management is core.** LibrePhotos isn’t just a viewer; editing, adding, moving, and deleting are part of the expected workflow for many users (even if your use case is “viewer only”). 
- **“Missing Photos” is deliberate.** Files can go missing when (a) they’re moved/renamed with another program, or (b) storage/mounting is misconfigured (e.g., a drive isn’t mounted). In these cases we **keep the photo in the DB and mark it as _missing_**, because the file often comes back. During scans we also try to **relink by hash** if the file reappears elsewhere.

This change appears to:

- **Break the existing “Photo is missing” behavior.**
- **Modify user-controlled fields in unexpected ways.**
- **Risk widespread deletion** when “skip raw” is enabled—which is not good.

### Why this is a problem

- **Data-retention expectations.** Users expect photos to remain visible/manageable even if source files are temporarily unavailable (e.g., moved, renamed or an external drive is unmounted today and remounted tomorrow).
- **Accidental loss.** Automatically removing records risks losing faces, people links, captions, ratings, albums, and other metadata we **do not reliably write back** to the image or XMP yet.
  - **Important:** **Saving back metadata does not cover all attributes.** Some user data remains only in the DB today. Therefore, **deleting XMP files or DB rows does not guarantee that user data is preserved** in fact, it can permanently remove details we can’t reconstruct.
- **Semantic mismatch.** The **Removed** state is reserved for items the user explicitly chose to delete, which are later cleaned up by a cron job. Using it for “temporarily missing” files changes its meaning and user expectations.

## Expected behavior (what we want to keep)

- If a RAW/source file can’t be found after enabling **skip raw**, the photo should be **marked _missing_**, **not** removed.
- **XMP-only entries.** That’s an edge case that should be **marked _missing_** too, not deleted; the photo object may still carry user metadata we need to preserve.
- **No user metadata** (faces, captions, ratings, people, albums) should be changed or deleted automatically.
- **Actual deletion** from the database should only occur via **Library Dashboard → “Delete all missing photos”**, after explicit user confirmation.

## Path forward

- **Skip RAW should not purge your library.** The “skip raw” option should influence **what to ingest/show**, not whether existing DB entries should be deleted. Turning it on must **not** cascade into removing photos or metadata from the DB.
- If your goal is a clean, shareable gallery of finals:  A PR that adds grouping logic (RAW + JPG + XMP as one) and/or a “Show only developed images” filter would work better for all users. 

I **won’t merge** changes that delete photo objects when their “main” file disappears or is an XMP. That’s too destructive and breaks the missing-photo contract.

Nikoh77 commented:
I completely understand — after all, I’ve only just started exploring LibrePhotos’ internal logic, so it’s normal that I don’t fully grasp the purpose of some features yet.

The idea of creating a single image object that groups together RAW, JPG, JPEG, XMP, and other related files seems to be the best approach. However, before I start working on it, I’d like to confirm that you agree the only reliable way to determine whether a file should belong to an existing image object is by comparing its filename.

Additionally, if a file is later moved or temporarily unavailable, it would simply be marked as missing, correct?

Another point I’d like to clarify: how does the current processing for face detection, tagging, etc. work? Does it happen only on the main image of each object?
Would it make sense for this processing to occur only on the main image when an image object contains multiple related files (e.g., JPG, RAW, XMP)?

Finally, the main image would be the first one LibrePhotos finds for that image object. Should it always remain the same, regardless of whether it’s a RAW or a JPG file?

So the general functioning would be: no skip raw option because in case there are more versions of the same image this would not be displayed twice but only the main file would be seen.

Nikoh77 committed:
Merge branch 'LibrePhotos:dev' into skipraw

derneuere commented:
Yes, only looking at the filename is good enough for now. There are cases where you can link with metadata e.g. live photos, but these are edge cases we can add later.

Yes, if a file was later moved, it will be marked as missing. Checking if a file is missing usually happens during the scan: https://docs.librephotos.com/docs/user-guide/missing-photos

All machine learning is done from the big thumbnail of the main file. Using a thumbnail instead of originals ensures, that the mechanism always works.

If a user prefers a jpeg over raw or the other way around depends on the user. I think it is save to prefer jpeg over raw, as jpegs are usually "developed" and raws are not.

Yes exactly there would only be one element on the timeline :)
